benchmark.py

```diff
@@ -0,0 +1,90 @@
+"""
+This script measures asyncio and threads while receiving files from
+the Internet
+"""
+import argparse
+import asyncio
+import itertools
+import multiprocessing.pool
+import os
+import time
+import urllib.request
+
+import aiohttp
+
+# Total amount of data is 860 Mb for both cases
+FILES = ("10Mb.txt", "50Mb.txt", "100Mb.txt", "200Mb.txt", "500Mb.txt")
+SMALL_FILES = ("10Mb.txt",) * 86
+
+
+async def async_downloader(url):
+    """Asynchronous download worker
+
+    url: URL to download from
+    """
+    async with aiohttp.ClientSession() as session:
+        async with session.get(url) as response:
+            with open(os.devnull, "wb") as file:
+                while True:
+                    # the size of chunk is the same as in urlretrieve
+                    chunk = await response.content.read(8192)
+                    if not chunk:
+                        break
+                    file.write(chunk)
+
+
+def downloader(url):
+    """Download worker
+
+    url: URL to download from
+    """
+    urllib.request.urlretrieve(url, filename=os.devnull)
+
+
+async def bench_asyncio(urls):
+    """Asynchronous benchmark
+
+    urls: list of URLs to download
+    """
+    coroutines = map(async_downloader, urls)
+    tasks = map(asyncio.create_task, coroutines)
+    await asyncio.gather(*tasks)
+
+
+def bench_threads(urls):
+    """Synchronous benchmark
+
+    urls: list to URLs to download
+    """
+    # There are 21 threads on my machine with 4 CPUs while running
+    # asyncio. I discovered experimentally that a pool with 17
+    # processes leads to the same behaviour.
+    with multiprocessing.pool.ThreadPool(17) as pool:
+        pool.map(downloader, urls)
+
+
+if __name__ == '__main__':
+    PARSER = argparse.ArgumentParser()
+
+    PARSER.add_argument("type",
+                        choices=("asyncio", "threads"),
+                        help="the way to download with")
+    PARSER.add_argument("--small-files",
+                        action="store_true",
+                        help="use small files",
+                        dest="small_files")
+
+    ARGS = PARSER.parse_args()
+
+    URLS = map(os.path.join,
+               itertools.repeat("http://ftp.mgts.by/test/"),
+               SMALL_FILES if ARGS.small_files else FILES)
+
+    if ARGS.type == "asyncio":
+        START_TIME = time.time()
```

Comment:
START_TIME definition can be moved below

Reply:
I'm not sure. It looks for me a little bit unfair, because the first case will starts after single `==` invocation, the second one - after 2 `==`, and so forth.